### ISSUE TITLE
Multiple UDP sources show same output on all endpoints

### ISSUE THREAD

**Issue by FlyveHest** [bluenviron/mediamtx#2133]
## Which version are you using?

MediaMTX v0.23.7

## Which operating system are you using?

- [x] Linux amd64 standard
- [ ] Linux amd64 Docker
- [ ] Linux arm64 standard
- [ ] Linux arm64 Docker
- [ ] Linux arm7 standard
- [ ] Linux arm7 Docker
- [ ] Linux arm6 standard
- [ ] Linux arm6 Docker
- [ ] Windows amd64 standard
- [ ] Windows amd64 Docker (WSL backend)
- [ ] macOS amd64 standard
- [ ] macOS amd64 Docker
- [ ] Other (please describe)

## Describe the issue

I've setup the mediamtx paths as follows

```
paths:
  testudpa:
    source: udp://x.x.x.x:5500
  testudpb:
    source: udp://y.y.y.y:5500
```

On startup, the log shows the following

```
2023/08/01 07:56:06 INF [path testudpa] [udp source] started
2023/08/01 07:56:06 INF [path testudpb] [udp source] started
2023/08/01 07:56:06 DEB [path testudpa] [udp source] connecting
2023/08/01 07:56:06 DEB [path testudpb] [udp source] connecting
2023/08/01 07:56:07 INF [path testudpa] [udp source] ready: 1 track (H264)
2023/08/01 07:56:07 INF [path testudpb] [udp source] ready: 1 track (H264)
```

But, connecting to either endpoint shows the same stream, and its seemingly random which streams "wins" the race, so, if I restart mediamtx and reconnect to the same endpoint I just was connected to, it will show the "other" stream, and again, it will then be available on both a and b.

## Describe how to replicate the issue

1. Setup multiple UDP sources on different paths
2. Start mediamtx
3. Conclude that the same stream is available on both endpoints, even though the source is different

## Did you attach the server logs?

yes

## Did you attach a network dump?

no

**Comment by FlyveHest** [comment]
> Is it a suitable solution?

Maybe, but it isn't a solution for the sitaution I am experiencing.  My two sources are multicast, but different IPs (same port, though), and I am getting the same stream on both paths.

**Comment by FlyveHest** [comment]
What I have done currently is something like this

`runOnDemand: ffmpeg -re -i "udp://x.x.x.x:5500?localaddr=127.0.0.1" ...`

x.x.x.x is the multicast IP and 127.0.0.1 is the IP of the interface that is part of the multicast network.

It works, but it is a cleaner solution to just be able to specify a multicast IP as a source.

I might still end up using ffmpeg anyway though, as the audio stream doesn't seem to compatible (not an aac stream)

**Comment by aler9** [comment]
@FlyveHest support for additional codecs is another feature that I'm interesting into developing, especially in this moment when SRT support has been added to the server, and SRT works with MPEG-TS too - which codec is it?

**Comment by FlyveHest** [comment]
These are all the audio codecs used in the streams i'm working with

```
"codec_long_name": "AAC (Advanced Audio Coding)"
"codec_long_name": "AAC LATM (Advanced Audio Coding LATM syntax)"
"codec_long_name": "ATSC A/52A (AC-3)"
"codec_long_name": "ATSC A/52B (AC-3, E-AC-3)"
"codec_long_name": "MP2 (MPEG audio layer 2)"
```

**Comment by calanor** [comment]
I have a similar issue i my ubuntu machine.
In configuration I have a only one udp source
```
paths:
  testudp:
    source: udp://239.1.1.5:5000
```
but in the same machine I have multiple subscriptions from others apps to another multicasts addresses at the same port but diferent IP, for example:
```
srt-live-transmit udp://@239.1.1.1:5000 ......
srt-live-transmit udp://@239.1.1.2:5000 ......
srt-live-transmit udp://@239.1.1.3:5000 ......
srt-live-transmit udp://@239.1.1.4:5000 ......
```
In this scenario, when I play a rstp://<ip>/testudp  with vlc or others player, I see the content of one of any multicast running on the machine, sometimes plays the source udp://239.1.1.5:5000. In all cases the stream closes unexpectedly and the video has some errors.
If in this example I close all the srt-live-transmit instances, the source its played correctly with any errors.

It seems that multicast subscription is not successful, but debugging the udp_source.go it appears that the JoinGroup are called with the correct address, but the data received does not seem correct.

**Comment by calanor** [comment]
Debugging the problem, I have seen with the lsof command that net.ListenPacket(restrictNetwork("udp", addr.String()))
makes a listen of
```
mediamtx  31964 31971            user1  23u     IPv4          318535611       0t0        UDP *:5000
mediamtx  31964 31968            user1  23u     IPv4          318535611       0t0        UDP *:5000
mediamtx  31964 31969            user1  23u     IPv4          318535611       0t0        UDP *:5000
mediamtx  31964 31968            user1  23u     IPv4          318535611       0t0        UDP *:5000
mediamtx  31964 31969            user1  23u     IPv4          318535611       0t0        UDP *:5000
```

instead a listen of
`mediamtx  31964 31969            user1  23u     IPv4          318535611       0t0        UDP 239.1.1.5:5000`

I suppose that ListenPacket could be used for unicast or for multicast on a machine where there are no more multicast subscriptions to the same port.
I guess if the address is multicast you should use net.ListenMulticastUDP(...) instead of net.ListenPacket or using syscalls. I'm not sure about it.

@FlyveHest, this could also be the cause of your problem because making 2 subscriptions causes two listens at  *:5500 and both paths will surely end up taking the same subscription. 
Pleas execute:
`sudo lsof | grep ":5500" `
when a readers connects to two paths and and paste the result.

**Comment by aler9** [comment]
Fixed by #2332

**Comment by calanor** [comment]
@aler9 I tested your patch and fix the issue but the `net.ListenPacket` function continue to bound to 0.0.0.0:5000 and listen all multicast addresses with the same port. Then it seems to filter to process only packets from the specific IP in` multicast.readFrom` function. 
This works fine but on my machine where I have more than 50 subscriptions made by other applications at the same port, it will result in all multicast traffic to port 5000 delivered to that machine to be given to the `multicast.readFrom`.  

It is a known issue that I could [see here](https://github.com/golang/go/issues/34728) where a possible [patch](https://github.com/danielkucera/go/commit/0226690a9b95ddafc72225f9c74cded1d72f12f5) is proposed but I don't know much about Golang and I do not know how to apply it in your code.

**Comment by aler9** [comment]
@calanor you're perfectly right about the performance issue caused by listening on `0.0.0.0` (on Linux only, while on Windows there are no issues). The patch you linked can't be applied since it changes the standard library, but there's an equivalent code here that i might be able to test and add somewhere in the future:

https://github.com/golang/go/issues/34728#issuecomment-1386037203

**Comment by aler9** [comment]
@calanor the server now listens on a specific IP only (https://github.com/bluenviron/gortsplib/pull/417).

**Comment by calanor** [comment]
Fantastic! 
Now we need to add the dependency of the new version of gortsplib to medimtx.

**Comment by calanor** [comment]
@aler9, For multicast the SO_REUSEPORT its not necessari.  SO_REUSEADDR and SO_REUSEPORT are identically for multicast addresses. SO_REUSEADDR implies SO_REUSEPORT for all multicast addresses and the other way round..

Also, I don't know how to test the new code in mediamtx. Wouldn't it be necessary to update go.sum to use the new version of gortsplib?

**Comment by aler9** [comment]
> I updated to gortsplib/v4 v4.0.2-0.20230913205520-e1869a8557ed, work but it has strange behavior.

That's because the multicast implementation shipped in gortsplib is bidirectional, it allows to receive and write multicast packets from/to all interfaces. In order to write multicast packets to all interfaces, we have to create a socket for each interface. Writing multicast packets is not needed by UDP / MPEG-TS, but is needed by RTSP, and for maintainability reasons i prefer not to create a specific multicast implementation, but to use a common one, even if this means overhead, that in this case is limited to the allocation of a series of unused sockets.

**Comment by calanor** [comment]
@aler9, I have researched this behavior.
I made a program in Go and C with only three instructions (not counting error checking):
Basically In C:
```
  recv_s = socket(AF_INET, SOCK_DGRAM, 0)
  setsockopt(recv_s, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(yes))
  bind(recv_s, (struct sockaddr *) &mcast_group, sizeof(mcast_group)
```

In Go:
```
readSock, err := syscall.Socket(syscall.AF_INET, syscall.SOCK_DGRAM, 0)
syscall.SetsockoptInt(readSock, syscall.SOL_SOCKET, syscall.SO_REUSEADDR, 1)
syscall.Bind(readSock, &lsa)
```
Executing `strace -f -e trace=network <test prgram>` they both show exactly the same calls:
```
socket(AF_INET, SOCK_DGRAM, IPPROTO_IP) = 3
setsockopt(3, SOL_SOCKET, SO_REUSEADDR, [1], 4) = 0
bind(3, {sa_family=AF_INET, sin_port=htons(5000), sin_addr=inet_addr("239.192.1.1")}, 16) = 0
```
But looking at the processes you can see how the Go program generates 5 threads while the C program only shows one thread.
Each thread in Go is a copy of the main one and for that reason 5 copies of the same udp bind are shown in lsof. This seems to be the reason why programs in Go seem to open up many more connections than programs made in other languages.
Try a simple [Hello World in Go](https://go.dev/doc/tutorial/getting-started) and look at the processes with` ps -T <pid>` you will see that a minimum of 5 threads are generated.
I could see more information [here](https://stackoverflow.com/a/27603427).

**Comment by aler9** [comment]
fixed in v1.1.0

**Comment by calanor** [comment]
Thanks, good job
